src/content/symbol-store.js

```diff
@@ -14,70 +30,86 @@ export class SymbolStore {
+      });
+
+      // Once the symbol table comes in, store it in the database, but don't
+      // let that block the promise that we return to our caller.
+      symbolTablePromise.then(symbolTable => {
+        this._db.storeSymbolTable(debugName, breakpadId, symbolTable).then(() => {
+          this._requestedSymbolTables.delete(libid);
+        }, error => {
+          console.error(`Failed to store the symbol table for ${debugName} ${breakpadId} in the database:`, error);
+        });
       });
-      this._importingLibs.set(libid, importPromise);
-      return importPromise;
-    }).catch(error => {
-      this._failedRequests.add(libid);
-      throw error;
+
+      return symbolTablePromise;
     });
+    this._requestedSymbolTables.set(libid, symbolTablePromise);
+    return symbolTablePromise;
   }
 
   /**
    * Get the array of symbol addresses for the given library.
-   * @param  {object} lib A library object with the properties `debugName` and `breakpadId`.
-   * @return {Promise<Array<Number>>} A promise of the array of addresses.
+   * @param  {Library} lib A library object with the properties `debugName` and `breakpadId`.
+   * @return {Promise<number[]>} A promise of the array of addresses.
    */
-  getFuncAddressTableForLib(lib) {
-    return this._ensureLibraryIsInDB(lib).then(
-      libKey => this._db.getFuncAddressTableForLib(libKey));
+  getFuncAddressTableForLib(lib: Library): Promise<Uint32Array> {
+    return this._getSymbolTable(lib).then(([addrs]) => addrs);
   }
 
   /**
    * Get an array of symbol strings for the given symbol indices.
-   * @param  {Array<Number>} requestedAddressesIndices An array where each element is the index of the symbol whose string should be looked up.
-   * @param  {Object} lib A library object with the properties `debugName` and `breakpadId`.
-   * @return {Promise<Array<String>>} An promise array of strings, in the order as requested.
+   * @param  {number[]} requestedAddressesIndices An array where each element is the index of the symbol whose string should be looked up.
+   * @param  {Library} lib A library object with the properties `debugName` and `breakpadId`.
+   * @return {Promise<string[]>} An promise array of strings, in the order as requested.
    */
-  getSymbolsForAddressesInLib(requestedAddressesIndices, lib) {
-    return this._ensureLibraryIsInDB(lib).then(
-      libKey => this._db.getSymbolsForAddressesInLib(requestedAddressesIndices, libKey));
+  getSymbolsForAddressesInLib(requestedAddressesIndices: number[], lib: Library): Promise<string[]> {
+    return this._getSymbolTable(lib).then(([, index, buffer]) => {
+      const decoder = new TextDecoder();
+      return requestedAddressesIndices.map(addrIndex => {
+        const startOffset = index[addrIndex];
+        const endOffset = index[addrIndex + 1];
+        const subarray = buffer.subarray(startOffset, endOffset);
+        return decoder.decode(subarray);
+      });
+    });
   }
 }
```

Comment:
I love the simplification of moving some complex logic out of the DB code to `symbol-store.js`